dts/bindings/pinctrl/st,stm32-pinctrl.yaml

```diff
@@ -0,0 +1,88 @@
+# Copyright (c) 2020 Linaro Limited
+# SPDX-License-Identifier: Apache-2.0
+
+description: STM32 PINCTRL
+
+compatible: "st,stm32-pinctrl"
+
+include: [base.yaml]
+
+properties:
+    reg:
+      required: true
+
+child-binding:
+    description: |
+        This binding gives a base representation of the STM32 pins configration
+
+    properties:
+        pinmux:
+          required: true
+          type: int
+          description: |
+            Adapted from https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/pinctrl/st,stm32-pinctrl.yaml
+            Integer array, represents gpio pin number and mux setting.
+            These defines are calculated as: ((port * 16 + line) << 8) | function
+            With:
+            - port: The gpio port index (PA = 0, PB = 1, ..., PK = 11)
+            - line: The line offset within the port (PA0 = 0, PA1 = 1, ..., PA15 = 15)
+            - function: The function number, can be:
+            * 0 : Alternate Function 0
+            * 1 : Alternate Function 1
+            * 2 : Alternate Function 2
+            * ...
+            * 15 : Alternate Function 15
+            * 16 : Analog
```

Comment:
We can add a 'GPIO' function here as well indeed.
But I wonder if this is really required. There's already a way to define GPIOs.
Do you have example of need for GPIO function that would be strictly different from existing gpio bindings and specific t pinctrl ?

Reply:
According to the Pinctrl DT spec you have to specify either pins, group, or pinmux in a pin configuration node. As Zephyr will for sure not implement pins and group now, somehow pinmux is required in a pin configuration node.

I see these solutions:
1) no pin configuration nodes for pins muxed to GPIO
2) deviate from the required pinmux
3) add  gpio input and gpio output to the selectable muxes.

As you can configure more than just the mux, solution 1) does not look so nice to me. Solution 2) is in violation to the DT spec. but works. In my special case I cant´t use these configuration nodes because my Pinctrl/edtlib mimic will issue an error on a missing pinmux/pins/group. Obviously I prefer solution 3).

As you explicitly stated that this will be no pinctrl driver work I think you are free to go with any solution.

Reply:
@b0661 sorry for high latency response. Just to be sure I'm getting your point:

> add gpio input and gpio output to the selectable muxes.

Do you expect something like :
`pinmux = <STM32_PINMUX('A', 9, GPIO_OUT/GPIO_IN)>`

Seems in Linux, they're using:
```
pinmux = <STM32_PINMUX('A', 9, GPIO)>
output-low/high; (No 'output-` is implicit `input`)
``` 

> According to the Pinctrl DT spec you have to specify either pins, group, or pinmux in a pin configuration node

For my information, do you have a pointer to this statement ?

Reply:
| Would it work in your case ?

Yes.

| For my information, do you have a pointer to this statement ?

The statement is from Linux (not the DT Spec.): https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pinmux-node.yaml

```
pinmux:
    description:
      The list of numeric pin ids and their mux settings that properties in the
      node apply to (either this, "pins" or "groups" have to be specified)
```

After reading again https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pinctrl-bindings.txt https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pincfg-node.yaml and https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pinmux-node.yaml I´m again puzzled, so don´ t take my view for the final truth. In my view you need to specify the pin(s) to apply a configuration to. So in the end you need one of pinmux, pins, groups, pinctrl-pin-array to specify the pin(s) of a pin configuration node. 

As long as Zephyr does not understand  pins, groups, pinctrl-pin-array  you have to use pinmux to specify the pin to apply e.g. bias to. As pinmux needs the mux function specified you need the GPIO selector in case you want to set the bias on a GPIO function type pin.

Reply:
> The statement is from Linux (not the DT Spec.) [...]

On my side, i'll retain the following sentence (from https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pinmux-node.yaml)
`While not required to be used, there are 3 generic forms of pin muxing nodes  which pin controller devices can use.`

>>  Would it work in your case ?
> Yes.

Good. Question is do we actually need this?
There's already a GPIO binding to define a GPIO function type pin. What will be the interest of having this alternate way?
On my side, I'd be affraid it could create confusion. At least in first step, I'd prefer to stay with the existing differentiation.
Then, maybe, once the need is clearly identified (as a lack) and people are familiar with new way of defining pinmux through device tree